src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java

```diff
@@ -1194,15 +1276,151 @@ private static void bipartitionSSTables(List<CompactionSSTable> sstables,
+         * CD, ABC or ABD.
+         * Also, the subset is selected by max timestamp order, oldest first, to avoid violating sstable time order. In
+         * the example above, if B is oldest and C is older than D, the two-input choice would be ABC (if A is older
+         * than D) or BC (if A is younger, avoiding combining C with A skipping D).
+         *
+         * @param pending A list where any "pending" compactions are placed, i.e. operations that will need to be
+         *                executed in the future.
+         * @param controller The compaction controller.
+         * @param spaceAvailable The amount of space available for compaction, limits the maximum number of sstables
+         *                       that can be selected.
+         * @return A compaction pick to execute next.
+         */
+        CompactionPick constructPicks(List<CompactionPick> pending, Controller controller, long spaceAvailable)
+        {
+            int count = maxOverlap;
+            int threshold = level.threshold;
+            int fanout = level.fanout;
+            int index = level.index;
+            int maxSSTablesToCompact = Math.max(fanout, (int) Math.min(spaceAvailable / level.avg, controller.maxSSTablesToCompact()));
+
+            assert count >= threshold;
+            if (count <= fanout)
+            {
+                /**
+                 * Happy path. We are not late or (for levelled) we are only so late that a compaction now will
+                 * have the  same effect as doing levelled compactions one by one. Compact all. We do not cap
+                 * this pick at maxSSTablesToCompact due to an assumption that maxSSTablesToCompact is much
+                 * greater than F. See {@link Controller#MAX_SSTABLES_TO_COMPACT_OPTION} for more details.
+                 */
+                return CompactionPick.create(index, allSSTablesSorted);
+            }
+            // The choices below assume that pulling the oldest sstables will reduce maxOverlap by the selected
+            // number of sstables. This is not always true (we may, e.g. select alternately from different overlap
+            // sections if the structure is complex enough), but is good enough heuristic that results in usable
+            // compaction sets.
+            else if (count <= fanout * controller.getFanout(index + 1) || maxSSTablesToCompact == fanout)
+            {
+                // Compaction is a bit late, but not enough to jump levels via layout compactions. We need a special
+                // case to cap compaction pick at maxSSTablesToCompact.
+                if (count <= maxSSTablesToCompact)
+                    return CompactionPick.create(index, allSSTablesSorted);
+
+                CompactionPick pick = CompactionPick.create(index, pullOldestSSTables(maxSSTablesToCompact));
+                count -= maxSSTablesToCompact;
```

Comment:
iiuc this is not accurate because we might have pulled less than `maxSSTablesToCompact`, is that ok?